gensim/sklearn_api/d2vmodel.py

```diff
@@ -164,7 +159,11 @@ def fit(self, X, y=None):
             The trained model.
 
         """
-        if isinstance(self._get_first(X), doc2vec.TaggedDocument):
+        def _get_first(some_iterable):
+            """Return first element of some iterable."""
+            for elem in some_iterable:
+                return elem
+        if isinstance(_get_first(X[:1]), doc2vec.TaggedDocument):
```

Comment:
Isn't this the same as `_get_first(_get_first(X))`?

I'm still confused by what `X` actually is, and what Pandas is doing. At the very least, this line deserves a thorough comment (though a clear code is always preferable).

Reply:
I thought that @mpenkov suggested `[:1]` part to improve performance. Since it part is confusing, I will remove it, because to me it seems redundant.

Reply:
I suggested it because it seemed correct, not because of performance.

If you look at the source before the change, it's:

```python
if isinstance([i for i in X[:1]][0], doc2vec.TaggedDocument):
    ...
```

My understanding of what that source is doing:

```python
tmp = X[:1]  # it'd be helpful to know what is X (and what is tmp) here
first = _get_first(tmp)  # first = [i for i in tmp][0]
if isinstance(first, doc2vec.TaggedDocument):
    ...
```

Is my understanding incorrect?

Reply:
Wow, I missed that. It should indeed be `_get_first(_get_first(X))`

Reply:
Can we somehow make it more obvious what X is in this case? It's the elephant in the room.

What data type is it?

Reply:
Finally, we also want a unit test to cover this code path, in order to prevent regressions.